routing/result_interpretation.go

```diff
@@ -431,6 +431,70 @@ func (i *interpretedResult) processPaymentOutcomeIntermediate(
 	case *lnwire.FailExpiryTooSoon:
 		reportAll()
 
+	// We only expect to get FailInvalidBlinding from an introduction node
+	// in a blinded route. The introduction node in a blinded route is
+	// always responsible for reporting errors for the blinded portion of
+	// the route (to protect the privacy of the members of the route), so
+	// we need to be careful not to unfairly "shoot the messenger".
+	//
+	// The introduction node has no incentive to falsely report errors to
+	// sabotage the blinded route because:
+	//   1. Its ability to route this payment is strictly tied to the
+	//      blinded route.
+	//   2. The pubkeys in the blinded route are ephemeral, so doing so
+	//      will have no impact on the nodes beyond the individual payment.
+	//
+	// Here we handle a few cases where we could unexpectedly receive this
+	// error:
+	// 1. Outside of a blinded route: erring node is not spec compliant.
+	// 2. Before the introduction point: erring node is not spec compliant.
+	//
+	// Note that we expect the case where this error is sent from a node
+	// after the introduction node to be handled elsewhere as this is part
+	// of a more general class of errors where the introduction node has
+	// failed to convert errors for the blinded route.
+	case *lnwire.FailInvalidBlinding:
+		introIdx, isBlinded := introductionPointIndex(route)
+
+		// Deal with cases where a node has incorrectly returned a
+		// blinding error:
+		// 1. A node before the introduction point returned it.
+		// 2. A node in a non-blinded route returned it.
+		if errorSourceIdx < introIdx || !isBlinded {
+			reportNode()
+			return
+		}
+
+		// Otherwise, the error was at the introduction node. All
+		// nodes up until the introduction node forwarded correctly,
+		// so we award them as successful.
+		if introIdx >= 1 {
+			i.successPairRange(route, 0, introIdx-1)
+		}
+
+		// If the hop after the introduction node that sent us an
+		// error is the final recipient, then we finally fail the
+		// payment because the receiver has generated a blinded route
+		// that they're unable to use. We have this special case so
+		// that we don't penalize the introduction node, and there is
+		// no point in retrying the payment while LND only supports
+		// one blinded route per payment.
+		//
+		// Note that if LND is extended to support multiple blinded
+		// routes, this will terminate the payment without re-trying
+		// the other routes.
+		if errorSourceIdx == len(route.Hops)-1 {
+			i.finalFailureReason = &reasonError
```

Comment:
>  When we penalize the last hop anyways we will not be able to retry anything either so we would fail in the next iteration with noroute ?

This is laying the groundwork for when we have multiple blinded path, in which case there will be the option to retry over a different blinded path (or retry this one with a smaller MPP split). I think it's worth keeping this distinction in place for when we come back to it.

Reply:
Hmm the comment says we do this because we don't want to penalize the intro point: 
```
//We have this special case so
// that we don't penalize the introduction node, and there is
// no point in retrying the payment while LND only supports
// one blinded route per payment.
```
But wouldn't we only fail the pair if we leave out this case ? 
```
i.failPairBalance(route, len(route.Hops)-1)
```

Let's imagine we have 2 blinded paths in the future, and keep the above restriction that we fail the payment if the blinded route is of length 2:

Let's say we split the payment and send it via those 2 paths, if only one makes it but the other because of some missing liquidity does not make it to the end, the node will abort the MPP, and our if clause will fail the payment although we could have maybe checked another combination of Amount-Splitting ?

So I still not fully understand why we special case blinded_path == 2 case, maybe you could elaborate a bit more?

Reply:
Original discussion of this idea is [here](https://github.com/lightningnetwork/lnd/pull/8095#discussion_r1364033298) - does that add useful context?

Reply:
Great, that added the necessary background. Failing the outer edge intro => final_hop does indeed punish the introduction node. For bimodal you could switch that off (`routerrpc.bimodal.nodeweight=0`) but  it makes indeed sense to have this cause handled differently.